### ISSUE TITLE
Release SecureDrop Client 0.14.0

### ISSUE THREAD

**Issue by zenmonkeykstop** [freedomofpress/securedrop-client#2282]
This issue tracks the SecureDrop Client release [version]. It will be organized by:

- Release Manager: @zenmonkeykstop
- Deputy Release Manager: @rocodes

This release includes the following changes:
- Adds support for selecting and deleting multiple sources
- Moves status bar to bottom of window layout
More detailed information on changes is available via https://github.com/freedomofpress/securedrop-client/milestone/13?closed=1

SecureDrop maintainers and testers: As you QA this release, please report back your testing results as comments on this ticket. File GitHub issues for any problems found, tag them "QA: Release", and associate them with the release milestone for tracking (or ask a maintainer to do so).

## Test plan

Prerequisite: start the server with
- n > 30 sources
- one or more testers: at least one source with a >1 MB file
- and start the client with debug logging

### Basic multi-delete/select testing (https://github.com/freedomofpress/securedrop-client/pull/2252)
- [ ] Ctrl+click selects multiple disjointed sources-  the widget items are distinctly highlighted
- [ ] Shift+click selects range of sources
- [ ] Mouse drag on sourcelist selects range of sources
- [ ] When more than one source is selected, the widget items are distinctly highlighted and the "Multiple sources selected" context view replaces the Source Conversation view
- [ ] Clicking on a single new source selects that source and de-selects the previously selected range
- [ ] When sources are deselected and a single source is selected, the correct conversation view appears. (Easiest way to test: reply to a source with their nickname, do this for a few sources, then toggle between them, selecting one then multiple. Observe match between visible conversation and selected source in sourcelist).
- [ ] When multiple sources are selected then deleted, and a new source is selected, the correct conversation view appears. (Select multiple sources, delete selected. Click on some remaining sources, and observe correct conversation rendered. Same nickname trick as above may be helpful).   
- [ ] Selecting Delete Sources from the toolbar prompts a dialog that shows each source name confirming deletion. When large numbers of sources are selected, all names are visible and the dialog area is scrollable.
- [ ] Closing or canceling the delete sources dialog has no effect, and the sources are still selected
- [ ] Accepting the delete sources dialog shows individualized deletion animations for each selected source, as they are removed.
- [ ] Clicking the Delete Sources toolbar button with no sources selected pops up a dialog informing that no sources are selected
- [ ] DeleteSources button is not visible when logged out, although the empty toolbar remains visible.

 ### General regression testing and UI inspection
- [ ] Source Menu (3-dots overflow menu by a single source) functionality is unchanged, including deleting a single source from the source menu. Please smoketest the actions in the 3-dot menu (Print Transcript shows print transcript dialog; Delete Source or Conversation from 3-dot over menu launches DeleteSource dialog for that source.)  
- [ ] Ensure that successive clicks on different sources in the sourcelist render the correct expanded conversation in the conversation pane (right side).  (Easiest way to do this is to reply with the sources nickname so that you can readily match the conversation to the source)
- [ ] Toggling between views (nothing selected, multi selected, single source selected) renders correct view
- [ ] Destructive dialog options (Delete Source, Delete Conversation) have a hover/mouseover state (https://github.com/freedomofpress/securedrop-client/pull/2208)
- [ ] Status messages and the status bar are correctly rendered at the _bottom_ of the application (https://github.com/freedomofpress/securedrop-client/pull/2230)

### Download error cases
- [ ] Begin download of large file, then delete the source (at the server if need be) before the download completes so that there is time for a sync. Client does not crash; warning present in logs. (https://github.com/freedomofpress/securedrop-client/pull/2231)
- [ ] Force a (small) download to fail, eg by disconnecting the network. Client does not crash, and well-formed error notifying about download failure is present in logs (https://github.com/freedomofpress/securedrop-client/issues/2274)
- [ ] Note: https://github.com/freedomofpress/securedrop-client/issues/2232 is out of scope for this release and is not a regression of this release. 

## Release tasks

- [ ] Check if there are any security bug fixes waiting to be pulled into the RC
- [ ] Check if there are any translations:
    - [ ] pending merge into `main`
    - [ ] pending inclusion as a supported language in [`MANIFEST.in`](https://github.com/freedomofpress/securedrop-client/blob/main/MANIFEST.in)
- [ ] Update changelog
- [ ] Create test plan
- [ ] Refresh nightlies
- [ ] Begin formal QA using nightlies; refresh nightlies as needed
- [ ] Build production package in standard [build environment](https://github.com/freedomofpress/securedrop-builder/wiki/FAQ#how-do-i-create-a-local-environment-suitable-for-building-packages)
- [ ] Sign production package
- [ ] Perform final pre-flight testing using apt-qa.freedom.press
  - [ ] **Localization:** In a dispVM, change your locale (e.g.: `export LANG=es_ES.utf-8; dpkg-reconfigure locales`), run the Client, and confirm that the application is translated.
- [ ] Publish production package
- [ ] Publicize release via support channels

**Comment by zenmonkeykstop** [comment]
## Test plan **IN PROGRESS**

Prerequisite: start the server with
- n > 30 sources
- one or more testers: at least one source with a >1 MB file
- and start the client with debug logging

### Basic multi-delete/select testing (https://github.com/freedomofpress/securedrop-client/pull/2252)
- [ ] Ctrl+click selects multiple disjointed sources-  the widget items are distinctly highlighted
- [ ] Shift+click selects range of sources
- [ ] Mouse drag on sourcelist selects range of sources
- [ ] When more than one source is selected, the widget items are distinctly highlighted and the "Multiple sources selected" context view replaces the Source Conversation view
- [ ] Clicking on a single new source selects that source and de-selects the previously selected range
- [ ] When sources are deselected and a single source is selected, the correct conversation view appears. (Easiest way to test: reply to a source with their nickname, do this for a few sources, then toggle between them, selecting one then multiple. Observe match between visible conversation and selected source in sourcelist).
- [ ] When multiple sources are selected then deleted, and a new source is selected, the correct conversation view appears. (Select multiple sources, delete selected. Click on some remaining sources, and observe correct conversation rendered. Same nickname trick as above may be helpful).   
- [ ] Selecting Delete Sources from the toolbar prompts a dialog that shows each source name confirming deletion. When large numbers of sources are selected, all names are visible and the dialog area is scrollable.
- [ ] Closing or canceling the delete sources dialog has no effect, and the sources are still selected
- [ ] Accepting the delete sources dialog shows individualized deletion animations for each selected source, as they are removed.
- [ ] Clicking the Delete Sources toolbar button with no sources selected pops up a dialog informing that no sources are selected
- [ ] DeleteSources button is not visible when logged out, although the empty toolbar remains visible.

 ### General regression testing and UI inspection
- [ ] Source Menu (3-dots overflow menu by a single source) functionality is unchanged, including deleting a single source from the source menu. Please smoketest the actions in the 3-dot menu (Print Transcript shows print transcript dialog; Delete Source or Conversation from 3-dot over menu launches DeleteSource dialog for that source.)  
- [ ] Ensure that successive clicks on different sources in the sourcelist render the correct expanded conversation in the conversation pane (right side).  (Easiest way to do this is to reply with the sources nickname so that you can readily match the conversation to the source)
- [ ] Toggling between views (nothing selected, multi selected, single source selected) renders correct view
- [ ] Destructive dialog options (Delete Source, Delete Conversation) have a hover/mouseover state (https://github.com/freedomofpress/securedrop-client/pull/2208)
- [ ] Status messages and the status bar are correctly rendered at the _bottom_ of the application (https://github.com/freedomofpress/securedrop-client/pull/2230)

### Download error cases
- [ ] Begin download of large file, then delete the source (at the server if need be) before the download completes so that there is time for a sync. Client does not crash; warning present in logs. (https://github.com/freedomofpress/securedrop-client/pull/2231)
- [ ] Force a (small) download to fail, eg by disconnecting the network. Client does not crash, and well-formed error notifying about download failure is present in logs (https://github.com/freedomofpress/securedrop-client/issues/2274)
- [ ] Note: https://github.com/freedomofpress/securedrop-client/issues/2232 is out of scope for this release and is not a regression of this release.

**Comment by zenmonkeykstop** [comment]
> Added one more section [on my testing](https://github.com/freedomofpress/securedrop-client/issues/2282#issuecomment-2470195167):

(tracked in #2296)

**Comment by zenmonkeykstop** [comment]
Rc2 is available with fixes for https://github.com/freedomofpress/securedrop-client/issues/2296

**Comment by zenmonkeykstop** [comment]
?

> ✔️ RC2 Concluded without the detection of any new issues.
> 
> One thing to note is that if the newsroom has less than the minimum threshold of sources (30, I think), it won't show a confirmation dialog. This is semantically what we wanted. I'm just stating it explicitly that in this case journalists won't get a confirmation.

I don't think that's right, there should always be a confirmation dialog.

- if 1-32 sources selected, the dialog should list all names
- if > 32 sources selected, the dialog lists first 30, and "plus additional", and delayed button
- if all sources selected, dialog displays "Notice: All.." in addition to whatever rules above say.

**Comment by zenmonkeykstop** [comment]
## Test plan - 0.14.0-rc2

Prerequisite: start the server with
- n > 30 sources
- one or more testers: at least one source with a >1 MB file
- and start the client with debug logging

### Basic multi-delete/select testing (https://github.com/freedomofpress/securedrop-client/pull/2252)
- [ ] Ctrl+click selects multiple disjointed sources-  the widget items are distinctly highlighted
- [ ] Shift+click selects range of sources
- [ ] Mouse drag on sourcelist selects range of sources
- [ ] When more than one source is selected, the widget items are distinctly highlighted and the "Multiple sources selected" context view replaces the Source Conversation view
- [ ] Clicking on a single new source selects that source and de-selects the previously selected range
- [ ] When sources are deselected and a single source is selected, the correct conversation view appears. (Easiest way to test: reply to a source with their nickname, do this for a few sources, then toggle between them, selecting one then multiple. Observe match between visible conversation and selected source in sourcelist).
- [ ] When multiple sources are selected then deleted, and a new source is selected, the correct conversation view appears. (Select multiple sources, delete selected. Click on some remaining sources, and observe correct conversation rendered. Same nickname trick as above may be helpful).   
- [ ] Selecting Delete Sources from the toolbar prompts a dialog that shows each source name confirming deletion. When large numbers of sources are selected, all names are visible and the dialog area is scrollable.
- [ ] Closing or canceling the delete sources dialog has no effect, and the sources are still selected
- [ ] Accepting the delete sources dialog shows individualized deletion animations for each selected source, as they are removed.
- [ ] Clicking the Delete Sources toolbar button with no sources selected pops up a dialog informing that no sources are selected
- [ ] DeleteSources button is not visible when logged out, although the empty toolbar remains visible.

 ### General regression testing and UI inspection
- [ ] Source Menu (3-dots overflow menu by a single source) functionality is unchanged, including deleting a single source from the source menu. Please smoketest the actions in the 3-dot menu (Print Transcript shows print transcript dialog; Delete Source or Conversation from 3-dot over menu launches DeleteSource dialog for that source.)  
- [ ] Ensure that successive clicks on different sources in the sourcelist render the correct expanded conversation in the conversation pane (right side).  (Easiest way to do this is to reply with the sources nickname so that you can readily match the conversation to the source)
- [ ] Toggling between views (nothing selected, multi selected, single source selected) renders correct view
- [ ] Destructive dialog options (Delete Source, Delete Conversation) have a hover/mouseover state (https://github.com/freedomofpress/securedrop-client/pull/2208)
- [ ] Status messages and the status bar are correctly rendered at the _bottom_ of the application (https://github.com/freedomofpress/securedrop-client/pull/2230)

### Download error cases
- [ ] Begin download of large file, then delete the source (at the server if need be) before the download completes so that there is time for a sync. Client does not crash; warning present in logs. (https://github.com/freedomofpress/securedrop-client/pull/2231)
- [ ] Force a (small) download to fail, eg by disconnecting the network. Client does not crash, and well-formed error notifying about download failure is present in logs (https://github.com/freedomofpress/securedrop-client/issues/2274)
- [ ] Note: https://github.com/freedomofpress/securedrop-client/issues/2232 is out of scope for this release and is not a regression of this release.

**Comment by rocodes** [comment]
> Accepting the delete sources dialog shows individualized deletion animations for each selected source, as they are removed.FAIL - sources deleted without the individual deletion blurbs on the conversation view.

The row corresponding to the deleted source(s) in the SourceList should block and show the "deleting" animation (and if it doesn't that's a fail), but the ConversationPane on the right should _not_ display individual source conversations with the deletion animation. As currently implemented, it should stay on the "Multiple Sources Selected" view until deletion is finished, where it should then display "Select a source to ....." (or "Nothing here" if all sources were deleted).   If a deletion animation is required I suggest a single one over the Multiple Sources Selected view, to minimize UI flickering and complexity, as a followup PR

**Comment by zenmonkeykstop** [comment]
Ok in that case it's as expected, thanks for the clarification!

**Comment by rocodes** [comment]
# apt-qa pre-release testing
Basic/smoketest since there has been previous qa
- Scenario: upgrade (via upgrades in template, not gui updater)
- [x] successful package upgrades
- [x] successful basic functionality testing (login, sync, read, reply, download, export, print)
- [x] successful multi-delete testing

**Comment by zenmonkeykstop** [comment]
# fresh install preflight
All good
- [x] fresh install from apt-qa succeeded
- [x] basic functionality is working (export tested but not print)
- [x] multi-delete is working.
- [ ]
